### PR TITLE
generateChangelog optionally creates runOnChange=true and replaceIfExists=true for createView changes

### LINKED ISSUE #4634: generateChangelog should optionally create runOnChange=true and replaceIfExists=true for createView changes
### Search first

- [X] I searched and no similar issues were found

### Description

Using the combination of `runOnChange=true` and `replaceIfExists=true` with `createView` changes makes lots of sense, as the source code of a view can `git diff`'ed, `git merge`'d and `git blame`'d then, which is not possible when using (lots) of subsequent `dropView` + `createView` changes.

Unfortunately `generateChangelog` statically sets both flags to false always, so the user is forced to manually fix it each time.

The idea behind this issue is to introduce two new flags for `generateChangelog`:
* `--create-view-run-on-change=true` will set `runOnChange="true"` for each `changeSet` which contains *solely* `createView` changes. For backwards compatibility, the default is `false`.
* `--create-view-replace-if-exists=true` will set `replaceIfExists="true"` for each `createView` change. For backwards compatibility, the default is `false`.

### Steps To Reproduce

```sql
CREATE VIEW V AS SELECT 1
```
```bash
generateChangelog --create-view-run-on-change=true --create-view-replace-if-exists=true
```

### Expected/Desired Behavior

```xml
<changeSet ... runOnChange="true">
    <createView ... replaceIfExists="true" viewName="V">SELECT 1</createView>
</changeSet>
```

### Liquibase Version

master

### Database Vendor & Version

SAP SQL Anywhere 17.0.11

### Liquibase Integration

N/A

### Liquibase Extensions

N/A

### OS and/or Infrastructure Type/Provider

Windows 10 Pro (German)

### Additional Context

N/A

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by mkarg**
Closes #4634 

## Impact

- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [x] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 

## Description

This PR introduces two new options:
* `--run-on-change-types=createView,createProcedure` will set `runOnChange="true"` for each `changeSet` which contains *solely* `createView` and `createProcedure` changes. For backwards compatibility, the default is an empty list.
* `--replace-if-exists=createView,createProcedure` will set `replaceIfExists="true"` for each `createView` and `createProcedure` change. For backwards compatibility, the default is an empty list. 

## Things to be aware of

N/A

## Things to worry about

N/A

## Additional Context

N/A

### CODE REVIEW

**Comment by mkarg** [comment]
Converted to draft to further improve this PR.

File: liquibase-standard/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java [hunk 1]
```diff
@@ -836,6 +846,30 @@ public void setChangeSetPath(String changeSetPath) {
...
 
+    public void setChangeSetRunOnChangeTypes(final String[] runOnChangeTypes) {
+        changeSetRunOnChangeTypes = runOnChangeTypes;
+    }
+
+    protected String[] getChangeSetRunOnChangeTypes() {
+        return changeSetRunOnChangeTypes;
+    }
+
+    private boolean isContainedInRunOnChangeTypes(final Change change) {
```
Comment (MalloD12) [inline liquibase-standard/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java:857]: It seems this method is causing some tests to fail. I ran **GenerateChangeLogMySQLIntegrationTest** and some tests and cleanup spec method are failing with below NPE:

```
liquibase.exception.CommandExecutionException: liquibase.exception.DatabaseException: java.lang.NullPointerException

	at liquibase.command.CommandScope.execute(CommandScope.java:237)
	at liquibase.command.util.CommandUtil.runDropAll(CommandUtil.groovy:112)
	at liquibase.command.core.GenerateChangeLogMySQLIntegrationTest.Ensure generated changelog map JSON columns correctly(GenerateChangeLogMySQLIntegrationTest.groovy:113)
Caused by: liquibase.exception.DatabaseException: java.lang.NullPointerException
	at liquibase.command.core.DropAllCommandStep.run(DropAllCommandStep.java:75)
	at liquibase.command.CommandScope.execute(CommandScope.java:214)
	... 2 more
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:221)
	at java.base/java.util.Arrays$ArrayList.<init>(Arrays.java:4323)
	at java.base/java.util.Arrays.asList(Arrays.java:4310)
	at liquibase.diff.output.changelog.DiffToChangeLog.isContainedInRunOnChangeTypes(DiffToChangeLog.java:858)
	at liquibase.diff.output.changelog.DiffToChangeLog.lambda$addToChangeSets$15(DiffToChangeLog.java:776)
	at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.base/java.util.Spliterators$ArraySpliterator.tryAdvance(Spliterators.java:958)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
	at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:533)
	at liquibase.diff.output.changelog.DiffToChangeLog.addToChangeSets(DiffToChangeLog.java:776)
	at liquibase.diff.output.changelog.DiffToChangeLog.generateChangeSets(DiffToChangeLog.java:327)
	at liquibase.database.AbstractJdbcDatabase.dropDatabaseObjects(AbstractJdbcDatabase.java:815)
	at liquibase.command.core.DropAllCommandStep.run(DropAllCommandStep.java:65)
	... 3 more
```
Could you please take a look at this?

File: liquibase-standard/src/main/java/liquibase/command/core/GenerateChangelogCommandStep.java [hunk 2]
```diff
@@ -68,6 +76,12 @@ public class GenerateChangelogCommandStep extends AbstractCommandStep {
                 .description("Directory to write table data to").build();
         OVERWRITE_OUTPUT_FILE_ARG = builder.argument("overwriteOutputFile", Boolean.class)
                 .defaultValue(false).description("Flag to allow overwriting of output changelog file. Default: false").build();
+        RUNONCHANGE_TYPES_ARG = builder.argument("runOnChangeTypes", String.class)
+                .defaultValue("none").description("Sets runOnChange=\"true\" for changesets containing solely changes of these types (e. g. createView, createProcedure, ...).").build();
+        final String replaceIfExistsTypeNames = supportedReplaceIfExistsTypes().collect(Collectors.joining(", "));
+        REPLACEIFEXISTS_TYPES_ARG = builder.argument("replaceIfExistsTypes", String.class)
+                .defaultValue("none")
+                .description(String.format("Sets replaceIfExists=\"true\" for changes of these types (supported types: createView, createProcedure)", replaceIfExistsTypeNames)).build();
```
Comment (MalloD12) [inline liquibase-standard/src/main/java/liquibase/command/core/GenerateChangelogCommandStep.java:84]: I think we are missing the '%s' on the provided text here.

**Comment by MalloD12** [comment]
Hi @mkarg, 

I left two review comments (one it's a very minor thing, though). Could you please look at them?

I was planning to add some integration test to do something like this:

```
def "Ensure generate changelog set runOnChange and replaceIfExists properties correctly"() {
        given:
        CommandUtil.runUpdate(mysql,"src/test/resources/changelogs/mysql/complete/view.procedure.changelog.xml")

        CommandScope commandScope = new CommandScope(GenerateChangelogCommandStep.COMMAND_NAME)
        commandScope.addArgumentValue(DbUrlConnectionCommandStep.URL_ARG, mysql.getConnectionUrl())
        commandScope.addArgumentValue(DbUrlConnectionCommandStep.USERNAME_ARG, mysql.getUsername())
        commandScope.addArgumentValue(DbUrlConnectionCommandStep.PASSWORD_ARG, mysql.getPassword())
        commandScope.addArgumentValue(GenerateChangelogCommandStep.REPLACEIFEXISTS_TYPES_ARG, "createView")
        commandScope.addArgumentValue(GenerateChangelogCommandStep.RUNONCHANGE_TYPES_ARG, "createView")

        OutputStream outputStream = new ByteArrayOutputStream()
        commandScope.setOutput(outputStream)

        when:
        commandScope.execute()

        then:
        def outputContent = outputStream.toString();
        outputContent.contains(" runOnChange=\"true\">")
        outputContent.contains(" replaceIfExists=\"true\"")

        cleanup:
        CommandUtil.runDropAll(mysql)
}
```
for which probably better to use `String.matches()` and use a regular expression to look for the `runOnChange` on the `changeset` tag and `replaceIfExists` in the `createView` (or createProcedure) tag.

**Comment by mkarg** [comment]
@MalloD12 I am PTO this week and look into this next week.

**Comment by MalloD12** [comment]
Hi @mkarg,

Have you had a chance to look at my comments?

Thanks,
Daniel.

**Comment by mkarg** [comment]
Sorry this week was booked out. Will try to find time for it next week.

**Comment by MalloD12** [comment]
Moving this PR to "In Discussion" column until we can resolve the above discussion.

File: liquibase-standard/src/main/java/liquibase/command/core/GenerateChangelogCommandStep.java (hunk 2 above)
Comment (mkarg) [reply liquibase-standard/src/main/java/liquibase/command/core/GenerateChangelogCommandStep.java]: You are right, I have just corrected this in 0e959aea1. Thank you for reporting it!

File: liquibase-standard/src/main/java/liquibase/command/core/GenerateChangelogCommandStep.java [hunk 3]
```diff
@@ -68,6 +76,12 @@
...
```
[the remark this answers is not in the document]
Comment (mkarg) [reply liquibase-standard/src/main/java/liquibase/command/core/GenerateChangelogCommandStep.java]: I just fixed this in 0e959aea1.

File: liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java [hunk 4]
```diff
@@ -26,7 +26,7 @@
 import java.util.Map;
 
 @DatabaseChange(name = "createProcedure", description = "Defines a stored procedure.", priority = ChangeMetaData.PRIORITY_DEFAULT)
-public class CreateProcedureChange extends AbstractChange implements DbmsTargetedChange {
+public class CreateProcedureChange extends AbstractChange implements DbmsTargetedChange, ReplaceIfExists {
```
[the remark this answers is not in the document]
Comment (mkarg) [reply liquibase-standard/src/main/java/liquibase/change/core/CreateProcedureChange.java:30]: This seems to be unrelated to my changes.

File: liquibase-standard/src/main/java/liquibase/change/core/CreateViewChange.java [hunk 5]
```diff
@@ -40,7 +40,7 @@
  * Creates a new view.
  */
 @DatabaseChange(name = "createView", description = "Create a new database view", priority = ChangeMetaData.PRIORITY_DEFAULT)
-public class CreateViewChange extends AbstractChange {
+public class CreateViewChange extends AbstractChange implements ReplaceIfExists {
```
[the remark this answers is not in the document]
Comment (mkarg) [reply liquibase-standard/src/main/java/liquibase/change/core/CreateViewChange.java:43]: This seems to be unrelated to my changes.

File: liquibase-standard/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java (hunk 1 above)
Comment (mkarg) [reply liquibase-standard/src/main/java/liquibase/diff/output/changelog/DiffToChangeLog.java:857]: Thank you for reporting this! Indeed I forgot to perfroam a `null` check. I have just fixed that in c87d9eac5e3d0b2d16a3162a6468846436c2af52.

**Comment by mkarg** [comment]
> I left two review comments (one it's a very minor thing, though). Could you please look at them?

Thank you for your patience and for reporting my faults. I fixed all of them (so I hope). :-)

> I was planning to add some integration test to do something like this:

Nice, thank you! :-)

> for which probably better to use `String.matches()` and use a regular expression to look for the `runOnChange` on the `changeset` tag and `replaceIfExists` in the `createView` (or createProcedure) tag.

I have no strong feeling about that, but sometimes simpler is better (easier to read).

**Comment by mkarg** [comment]
> Moving this PR to "In Discussion" column until we can resolve the above discussion.

IMHO discussion is resolved now.

**Review by MalloD12 — APPROVED** [review]
**Approved**.

## Review and testing results:

Code changes looks good to me. Build and most of the test checks (except Test harness for a known issue) have been successfully executed. I manually test this change and looks good to me, but I also fixed an existent command definition test, plus added an integration test to validate this change as well (only did it for createView change, since createProcedure is only generate in changelog for our PRO version).

Note: "CodeQL Analyze" check failed, but it's also for known issue which devops is already working on. I created a copy of this PR and I could  see that check successfully executed. 

### Things to be aware of:
- None: at the moment this is only applicable for `createView` and `createProcedure` change types.

### Things to worry about:
- None

**Review by filipelautert — APPROVED** [review]

**Comment by MalloD12** [comment]
@mkarg FYI I have updated your code a bit as requested to extend the behavior you added here also for `Diff-Changelog` command (which's similar to `generate-changelog`). If you have any objection/concerns please let me know.

@filipelautert could you please review this PR again when you have a chance?

@rberezen it's ready for you to test when you have a chance.

Thanks,
Daniel.

**Comment by mkarg** [comment]
> @mkarg FYI I have updated your code a bit as requested to extend the behavior you added here also for `Diff-Changelog` command (which's similar to `generate-changelog`). If you have any objection/concerns please let me know.

As long as it does not delay the integration of this PR any further I do not have any objections.

**Comment by tati-qalified** [comment]
I've manually tested this PR, using both `--replace-if-exists-types=createView,createProcedure` and `--run-on-change-types=createView,createProcedure` and all combinations of the following:

- Database
     - PostgreSQL
     - MySQL
     - DB2
     - Oracle
- Changelog file type
     - xml
     - yml
     - sql
     - json  
- Command
     - `generate-changelog`
     - `diff-changelog`  

The point of this was to try being as thorough as possible, with some of the most widely used databases.

### **Here are the results:**
It works as expected for file types xml, yml and json, for all 4 tested databases and both commands.

There are some problems with the sql files for all 4 databases. I'll specify them below.

**All databases**
None of the changesets are decorated with `runOnChange=true`, both for `generate-changelog` and for `diff-changelog`

**PostgreSQL, Oracle, MySQL**
The `create procedure` changesets are created as follows (with the correct SQL syntax for each database):
```
-- changeset tfernandez:1702560678224-5 splitStatements:true
DROP PROCEDURE IF EXISTS "procedure1";
CREATE OR REPLACE PROCEDURE public.procedure1()
 LANGUAGE plpgsql
AS $procedure$
	BEGIN
		insert into public.entities (id, name) values (1, 'two');
	END;
$procedure$;
```
The statement `DROP PROCEDURE IF EXISTS` shouldn't be there. This is true when running either command.


**Oracle**
Running `generate-changelog` generates the `CREATE VIEW` changeset as `CREATE OR REPLACE FORCE VIEW "NEWVIEW" ("A") AS SELECT 1 AS A FROM DUAL;`.
Running `diff-changelog` generates it as `CREATE OR REPLACE FORCE VIEW NEWVIEW (A) AS SELECT 1 AS A FROM DUAL;`.
Note that one uses quotes for the view name and the other one doesn't. Both statements are correct, but this behaviour should be consistent.

**PostgreSQL**
The `CREATE VIEW` changesets are generated as follows:
```
DROP VIEW IF EXISTS "newview";
CREATE VIEW "newview" AS SELECT id,  name
   FROM entities;
```
The `DROP VIEW IF EXISTS` statement shouldn't be there. This happens when running either command.


______________________________________________________________________________________________________________

That's all I found. Let me know if there is anything else I can help with.

**Comment by filipelautert** [comment]
@MalloD12 can you review @tati-qalified comments? thanks!

**Comment by MalloD12** [comment]
Thanks @tati-qalified for your very detailed testing. I think these "issues"/inconsistencies are not because of these changes.  For example, in the case of `runOnChange` property looking into `CreateViewGenerator` or `CreateProcedureGenerator` we can see none of them have support for `runOnChange`and I think that's why we don't see it when generating a SQL output.

Regarding the "drop procedure" statement you are right it shouldn't be there, but again this is an existent piece in the code I would recommend similarly to the previous finding to deal with in a separate issue.

I think the same applies to the other findings. 

Again, I would recommend dealing with each of these findings separately, so we don't keep this ticket on hold more time since these findings haven't been introduced with this PR.

What do you think @filipelautert - @rberezen - @tati-qalified?

Thanks,
Daniel.

**Comment by rberezen** [comment]
@MalloD12, I agree with you on creating a separate issue for the problems described

**Comment by tati-qalified** [comment]
Sounds good to me

**Review by filipelautert — APPROVED** [review]

**Comment by filipelautert** [comment]
@MalloD12 could you fix the conflicts?
@rberezen could you please finish the review?

**Comment by MalloD12** [comment]
> @MalloD12 could you fix the conflicts? @rberezen could you please finish the review?

@filipelautert done!

**Review by rberezen — APPROVED** [review]

**Comment by adrian-velonis1** [comment]
Internal documentation ticket: https://datical.atlassian.net/browse/PD-4098

**Pull request merged**